### PR TITLE
Add session identification to PII filter 

### PR DESCRIPTION
**Pull request by pavolloffay**
Notable changes:
* add session identification to PII filter
* adds parsed attribute struct that holds a flattened object (JSON, cookie)
* adds some comments to PII config

Follow up PRs
- [ ] remove processor helper to pass updated context to the next processor
- [ ] make creation of flat map store conditional based on an attribute from sampling processor   

My notes:
-PII config https://github.com/Traceableai/traceableai/blob/main/install/agent/kubernetes/traceableai-collector-only.yaml#L63

### CODE REVIEW

File: processors/parsed_attribte.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+// PutParsedAttribute puts ParsedAttribute into ParsedSpanData.
+func (p *ParsedSpanData) PutParsedAttribute(key string, attr *ParsedAttribute) {
+	p.parsedAttributes[key] = attr
+}
+
+// ParsedAttribute encapsulates parsed a single attribute.
+type ParsedAttribute struct {
+	// flattered JSON, cookie of not redacted fields
+	// e.g. JSON {"a": "b"} will contain "$.a": "b"
```
Comment (davexroth) [inline processors/parsed_attribte.go:67]: why do we need to prefix it with `$.` json path notation?

File: processors/parsed_attribte.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	ptd := &ParsedTracesData{
+		spanAttributeMap: map[pdata.SpanID]*ParsedSpanData{},
+	}
+	ctx = context.WithValue(ctx, contextKey{}, ptd)
+	return ctx, ptd
+}
+
+// ParsedTracesData
+type ParsedTracesData struct {
```
Comment (davexroth) [inline processors/parsed_attribte.go:25]: Is this a trace, or just a map of arbitrarily batched spans that's being processed?

File: processors/parsed_attribte.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+type contextKey struct{}
+
+// FromContext returns ParsedTracesData and updated context
+func FromContext(ctx context.Context) (context.Context, *ParsedTracesData) {
+	sAttr, ok := ctx.Value(contextKey{}).(*ParsedTracesData)
+	if ok {
```
Comment (jcchavezs) [inline processors/parsed_attribte.go:14]: Should we add a log for this lack of data?

File: processors/parsed_attribte.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// GetParsedSpanData returns ParsedSpanData for a given span.
+func (p *ParsedTracesData) GetParsedSpanData(spanID pdata.SpanID) *ParsedSpanData {
+	pSpanData, ok := p.spanAttributeMap[spanID]
+	if !ok {
+		pSpanData = &ParsedSpanData{
+			parsedAttributes: map[string]*ParsedAttribute{},
+		}
+		p.spanAttributeMap[spanID] = pSpanData
```
Comment (jcchavezs) [inline processors/parsed_attribte.go]: Is spanID enough? I'd rather use traceID+spanID as spanID is unique in a trace but not generally unique.

File: processors/piifilterprocessor/processor_test.go [hunk 5]
```diff
@@ -247,3 +249,49 @@ func assertJSONEqual(t *testing.T, expected, actual string) {
...
+	cfg := &piifilterprocessor.TransportConfig{
+		KeyRegExs: []piifilterprocessor.TransportPiiElement{
+			{RegexPattern: "^password$"},
+		},
+		RedactStrategyName: "hash",
+		ComplexData: []piifilterprocessor.TransportPiiComplexData{
+			{
+				Key:  "http.request.body",
+				Type: "json",
+				//TypeKey: "http.request.headers.content-type",
```
Comment (jcchavezs) [inline processors/piifilterprocessor/processor_test.go]:
```suggestion
```

File: processors/parsed_attribte.go (hunk 1 above)
Comment (jcchavezs) [reply processors/parsed_attribte.go:67]: I guess to differentiate from actual keys in flattened structure?

**Review by jcchavezs — APPROVED** [review]

File: processors/parsed_attribte.go (hunk 2 above)
Comment (pavolloffay) [reply processors/parsed_attribte.go:25]: multiple traces therefore `Traces`

Comment (pavolloffay) [reply processors/parsed_attribte.go:25]: Added more comments

File: processors/parsed_attribte.go (hunk 3 above)
Comment (pavolloffay) [reply processors/parsed_attribte.go:14]: The API is used even when creating the parsed data. If we split it into create and get then the get should maybe log warning.

File: processors/parsed_attribte.go (hunk 1 above)
Comment (pavolloffay) [reply processors/parsed_attribte.go:67]: I think the modsec requires this format https://github.com/Traceableai/platform-gateway/blob/main/api-utils/src/test/java/ai/traceable/platform/apiutils/SpanApiUtilsTest.java#L1645

File: processors/parsed_attribte.go (hunk 4 above)
Comment (pavolloffay) [reply processors/parsed_attribte.go]: There is definitely some probability to have a collision, I have changed it to Span struct as key - it's even more idiomatic.

**Comment by pavolloffay** [comment]
I will merge on green feel free to comment and I will apply fix in a follow-up PRs.

**Pull request merged**